src/sage/rings/finite_rings/element_base.pyx

```diff
@@ -97,18 +112,26 @@ cdef class FiniteRingElement(CommutativeRingElement):
             F = n.factor()
         from sage.groups.generic import discrete_log
         if algorithm is None or algorithm == 'Johnston':
-            g = K.multiplicative_generator()
+            # In the style of the Adleman-Manders-Miller algorithm,
+            # we will use small order elements instead of a multiplicative
+            # generator, which can be expensive to compute.
             for r, v in F:
+                # 0 < v <= k
                 k, h = (q-1).val_unit(r)
-                z = h * (-h).inverse_mod(r**v)
+                hinv = (-h).inverse_mod(r**v)
+                z = h * hinv
                 x = (1 + z) // r**v
-                if k == 1:
+                if k == v:
                     self = self**x
                 else:
-                    t = discrete_log(self**h, g**(r**v*h), r**(k-v), operation='*')
-                    self = self**x * g**(-z*t)
+                    # We need an element of order r^k (g^h in Johnston's article)
+                    # self^x differs from the actual nth root by an element of
+                    # order dividing r^(k-v)
+                    gh = K.zeta(r**k)
```

Comment:
Here is possibly a place where time can be saved even more. Namely computing `r**k` forgots the factored form of this number and is recomputed in the last line of `zeta`. Maybe one can tweak `zeta` so that it also accepts `Factorization` argument.